### ISSUE TITLE
Passing shared changing parameters to parallel envs during training.

### ISSUE THREAD

**Issue by yannidd** [astooke/rlpyt#181]
Hi, I am building an automatic domain randomisation (ADR) algorithm that samples environment parameters from an ever-growing distribution to make the environment harder as the agent learns. At the end of each episode, each of the parallel environments generates episode info, which includes a performance metric on the current environment parameters. This metric is used to shrink/grow the parameter distributions. However, all this is just theory for now and I am trying to figure out how it can be implemented in `rlpyt`. I saw a similar problem (#63), but it wasn't clear where the parameters come from.

My current idea is to modify the `collect_batch()` method of `CpuResetCollector`. I need to somehow pass an instance of my own ADR class to the collector, and inside the collector I would call the `sample_env_params()` and `update_distributions(performance_from_env)` methods. I think it should look something like this:

```python
    def collect_batch(self, agent_inputs, traj_infos, itr):
        ...
        for t in range(self.batch_T):
            ...
            for b, env in enumerate(self.envs):
                o, r, d, env_info = env.step(action[b])
                ...
                if getattr(env_info, "traj_done", d):
                    ...
                    adr.update_distributions(env_info['adr_performance'])
                    env.set_params(adr.sample_env_params())
                    o = env.reset()
                if d:
                    self.agent.reset_one(idx=b)
                ...
```

So, my questions are: 1) is this a good approach of doing this; 2) if so, what is a good way to pass the ADR class instance to the collector; 3) I am not sure how the parallel environments work, but it seems like their step() methods are called in a for loop - is the method suggested above thread-safe?

**Comment by astooke** [comment]
Interesting algorithm!  OK there are probably multiple ways to do this, let see...

I think you have a reasonable approach, to make an ADR object to handle this across multiple environments.

Here's a thought as to an easy way to parallelize:

1. Instantiate your ADR object in the train script, and include this object in the `env_kwargs` dict that goes into the sampler.  Then every environment will receive this object when it is instantiated, even inside worker processes.
2. In your environment, inside the `step()` method, when it reaches the end of a trajectory, send the info to the ADR object `adr.update_distro(adr_perf)`.
3. In your environment, inside the `reset()` method, have it first sample ADR params, and then do the rest of the reset.

The part that needs to be parallelized is the `adr.update_distro()` method.  You could start just with the `SerialSampler` and make sure the numbers are doing what you want.  Then you could parallelize by pre-allocating some shared memory buffer inside the ADR object when it's instantiated, and whenever you have new params, inside the `adr.update_distro()`, you'll write them like `adr.param_1[:] = update_func(adr.param_1, new_adr_perf_values)`, and you'll want to put a lock around that.

This way, you shouldn't have to change any of the collector or sampler or anything else.

And if you want to log ADR values on a trajectory-wise basis, you can make a `AdrTrajInfoCls` and have it grab those values in the `step()` method whenever `done=True`.

How does that sound?

**Comment by yannidd** [comment]
That's a great suggestion, I didn't think of getting most of the work handled by the ADR object and using shared memory with locks. It also sounds like the least amount of hassle since I don't have to change the RL framework. 

I'll work on this in the upcoming week so closing for now, but I'll reopen if any issues pop up :) P.S. Thanks for the quick reply.

**Comment by yannidd** [comment]
Hi, just a quick update on progress. I followed your recommendation exactly and used `multiprocessing.Value` and `multiprocessing.Array` to store the distribution information and its corresponding buffer with performances in an `AdrParamSet` object. I created an instance attribute `self.lock = multiprocessing.Lock()` and wrapped the update function with a context manager `with self.lock:` to sync between processes. All the functionality was implemented without touching the `rlpyt` code. 

I did a quick test on my custom env for 6M iterations and it all seems to work as expected - after about 3M iterations, the robot learns to walk and the distributions start expanding.

Thanks again for the suggestions and the great library (I spent a month trying out different RL libraries and rlpyt was the most performant and the easiest to implement custom RNN models).
